File: api/parse-pdf.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';
import * as pdfjsLib from 'pdfjs-dist/legacy/build/pdf.mjs';

interface RawRow {
  index: number;
  cells: string[];
}

interface TempItem {
  str: string;
  x: number;
  y: number;
  width: number;
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  // CORS configuration
  res.setHeader('Access-Control-Allow-Credentials', 'true');
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
  res.setHeader(
    'Access-Control-Allow-Headers',
    'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version'
  );

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method Not Allowed' });
  }

  const { fileBase64, password } = req.body || {};

  try {
    if (!fileBase64) {
      return res.status(400).json({ error: 'Missing fileBase64 data' });
    }

    const buffer = Buffer.from(fileBase64, 'base64');
    const loadingTask = pdfjsLib.getDocument({
      data: new Uint8Array(buffer),
      password: password || '',
      useWorkerFetch: false,
      isEvalSupported: false,
    });

    const pdfDoc = await loadingTask.promise;
    const rawRows: RawRow[] = [];
    let rowIndexCounter = 0;

    for (let pageNum = 1; pageNum <= pdfDoc.numPages; pageNum++) {
      const page = await pdfDoc.getPage(pageNum);
      const textContent = await page.getTextContent();

      const items: TempItem[] = textContent.items
        .map((item: any) => ({
          str: item.str,
          x: item.transform[4],
          y: item.transform[5],
          width: item.width || (item.str.length * 6),
        }))
        .filter(item => item.str.trim() !== '');

      if (items.length === 0) continue;

      const rowGroups: { y: number; items: TempItem[] }[] = [];

      items.forEach(item => {
        const foundGroup = rowGroups.find(g => Math.abs(g.y - item.y) < 10);
        if (foundGroup) {
          foundGroup.items.push(item);
        } else {
          rowGroups.push({ y: item.y, items: [item] });
        }
      });

      rowGroups.sort((a, b) => b.y - a.y);

      rowGroups.forEach(group => {
        group.items.sort((a, b) => a.x - b.x);

        const cells: string[] = [];
        let currentCell = '';
        let lastRight = -999;

        group.items.forEach(item => {
          if (lastRight === -999) {
            currentCell = item.str;
            lastRight = item.x + item.width;
          } else if (item.x - lastRight < 6) {
            currentCell += ' ' + item.str;
            lastRight = Math.max(lastRight, item.x + item.width);
          } else {
            cells.push(currentCell.trim());
            currentCell = item.str;
            lastRight = item.x + item.width;
          }
        });

        if (currentCell) {
          cells.push(currentCell.trim());
        }

        if (cells.length > 0) {
          rawRows.push({
            index: rowIndexCounter++,
            cells,
          });
        }
      });
    }

    return res.status(200).json({ rawRows });
  } catch (error: any) {
    console.error('Server PDF parse error:', error);
    const msg = error.message || '';
    if (error.name === 'PasswordException' || msg.toLowerCase().includes('password')) {
      return res.status(401).json({ error: password ? 'INCORRECT_PASSWORD' : 'PASSWORD_REQUIRED' });
    }
    return res.status(500).json({ error: 'Failed to parse PDF: ' + msg });
  }
}

File: api/parse-pdf.py
from http.server import BaseHTTPRequestHandler
import json
import base64
import io
from pypdf import PdfReader
from pypdf.errors import WrongPasswordError, FileNotDecryptedError

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS, POST')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-CSRF-Token, X-Requested-With, Accept')
        self.end_headers()

    def do_POST(self):
        password = ''
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            if content_length == 0:
                self._send_json(400, {'error': 'Empty request body'})
                return

            body_bytes = self.rfile.read(content_length)
            body = json.loads(body_bytes.decode('utf-8'))
            
            file_base64 = body.get('fileBase64', '')
            password = body.get('password', '')

            if not file_base64:
                self._send_json(400, {'error': 'Missing fileBase64 data'})
                return

            pdf_bytes = base64.b64decode(file_base64)
            pdf_stream = io.BytesIO(pdf_bytes)
            
            reader = PdfReader(pdf_stream)
            
            if reader.is_encrypted:
                decrypted = False
                if password:
                    try:
                        decrypted = (reader.decrypt(password) != 0)
                    except Exception:
                        decrypted = False
                
                if not decrypted:
                    error_msg = 'INCORRECT_PASSWORD' if password else 'PASSWORD_REQUIRED'
                    self._send_json(401, {'error': error_msg})
                    return

            raw_rows = []
            row_index_counter = 0

            for page_idx, page in enumerate(reader.pages):
                items = []

                def visitor_body(text, cm, tm, font_dict, font_size):
                    if text and text.strip():
                        # tm[4] is X coordinate, tm[5] is Y coordinate
                        items.append({
                            'str': text,
                            'x': float(tm[4]),
                            'y': float(tm[5]),
                            'width': float(len(text) * (font_size or 10) * 0.5)
                        })

                page.extract_text(visitor_text=visitor_body)

                if not items:
                    continue

                # Group items into rows by Y coordinate (< 10px difference)
                row_groups = []
                for item in items:
                    found = False
                    for g in row_groups:
                        if abs(g['y'] - item['y']) < 10:
                            g['items'].append(item)
                            found = True
                            break
                    if not found:
                        row_groups.append({'y': item['y'], 'items': [item]})

                # Sort rows top-to-bottom (Y descending)
                row_groups.sort(key=lambda g: g['y'], reverse=True)

                for group in row_groups:
                    # Sort items inside row left-to-right (X ascending)
                    group['items'].sort(key=lambda item: item['x'])

                    cells = []
                    current_cell = ""
                    last_right = -999.0

                    for item in group['items']:
                        if last_right == -999.0:
                            current_cell = item['str']
                            last_right = item['x'] + item['width']
                        elif item['x'] - last_right < 6:
                            current_cell += " " + item['str']
                            last_right = max(last_right, item['x'] + item['width'])
                        else:
                            if current_cell.strip():
                                cells.append(current_cell.strip())
                            current_cell = item['str']
                            last_right = item['x'] + item['width']

                    if current_cell.strip():
                        cells.append(current_cell.strip())

                    if cells:
                        raw_rows.append({
                            'index': row_index_counter,
                            'cells': cells
                        })
                        row_index_counter += 1

            self._send_json(200, {'rawRows': raw_rows})

        except (WrongPasswordError, FileNotDecryptedError):
            self._send_json(401, {'error': 'INCORRECT_PASSWORD' if password else 'PASSWORD_REQUIRED'})
        except Exception as e:
            self._send_json(500, {'error': f'Failed to parse PDF: {str(e)}'})

    def _send_json(self, status_code, data):
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS, POST')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-CSRF-Token, X-Requested-With, Accept')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode('utf-8'))
